patroni/postgresql.py

```diff
@@ -1645,52 +1717,77 @@ def basebackup(self, conn_url, env):
                      FROM pg_stat_replication
-                    ORDER BY flush_{0} DESC""".format(self.lsn_name)):
+                     ORDER BY flush_{0} DESC""".format(self.lsn_name)):
             member = members.get(app_name)
             if state != 'streaming' or not member or member.tags.get('nosync', False):
                 continue
-            if sync_state == 'sync':
-                return app_name, True
-            if sync_state == 'potential' and app_name == current:
-                # Prefer current even if not the best one any more to avoid indecisivness and spurious swaps.
-                return current, False
-            if sync_state == 'async':
-                candidates.append(app_name)
-
-        if candidates:
-            return candidates[0], False
-        return None, False
-
-    def set_synchronous_standby(self, name):
-        """Sets a node to be synchronous standby and if changed does a reload for PostgreSQL."""
-        if name and name != '*':
-            name = quote_ident(name)
-        if name != self._synchronous_standby_names:
-            if name is None:
+            if state == 'sync' or (state == 'potential' and current is None):
+                current = member.name
+            active.append(member.name)
+
+        result['active'] = set(active)
+        result['numsync'] = result['num'] + 1 if not result.get('has_star') else 1
+        result['sync'] = set([self.name] + result['members']).difference(['*'])
+
+        return result
+
+    def set_synchronous_state(self, num, sync=None):
+        """Updates synchronization state.
+
+        Parameters:
+        `num` specifies number of nodes in sync (including leader).
+        `sync` - set of nodes to sync to, including leader.
+
+        Number of nodes in sync can be bigger than the set to sync to. Commiters will then have to wait for a new
+        replica to become sync. As a special case with num >= 2 and no standbys, synchronous_standby_names will get
+        set to `*`. Such a state will only be enabled with a 1/1 quorum and minimum_replication_factor >= 2. The
+        state parses as 1/1 sync and if any node appears the sync state will first get changed. At worst any unknown
+        node being synced to will be unknown to Patroni and we will not be able to promote it on leader failure. This
+        can happen anyway in the case if one of the nodes is configured with nofailover, but without nosync.
+
+        TODO: Should we have a special synchronization mode that requires that all nodes that accept synchronization
+        must also be eligible for failover?
+        """
+
+        if num == 1 or num is None:
+            # Turn off sync replication
+            assert not sync or sync == set([self.name])
+            ssn = None
+        else:
+            assert self.name in sync
+            sync_standbys = sync.difference([self.name])
+            standby_list = ", ".join(sorted(sync_standbys)) if sync_standbys else "*"
```

Comment:
Lost quote_ident.

Reply:
Thanks for catching this. I will add it back in.